sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,390 @@
+
+2. Implict instead of explicit. 
+
+   a. The constuctor performs discovery depending on the parameter combination provided. A common example of this is when fetching credentials. The Client will search for credential details in the netrc only if a Keystore is not found. This is unnecessarily added work and as a result of this implicity, the user needs to be fully aware of the context of their environment.
+
+3. Areas of volatility are not encapsulated.
+
+   a. The constructor's parameters in the method signature are not encapsulated. It accepts 9 hanging parameters when they can instead be encapsulated into related objects. 
+
+   b. We can expect changes in the following areas in future iterations:
+
+   ​	i. Method of fetching configuration. Currently, we support fetching configuration details from the conjurrc file but this can change in the future and we will support methods in the future (via environment variables, etc).
+
+   ​	ii. Method of fetching credentials (via the netrc file or keystore). Currently, we support fetching credential data from the keystore or the netrc file.
+
+   Because of all these possible changes, we need to encapsulate them so that when changes arise, they are contained, without effecting the system.
+
+For these reasons, I propose we extract the configuration and credential information to their own objects to be created by the user before the Client is called. In those classes, all the collection and validations will be performed. As a result, the Client constructor will not be forced to be in charge of handle multiple responsibilities, the user will have to explicitly define the and the areas of volatility will be contained.
+
+We can handle the separation in one of the following ways:
+
+|      | Solution                                        | Description                                                  | Pros                                                         | Cons                                                         |
+| ---- | ----------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Keep as is                                      |                                                              | - Doesn't break current user apps<br />- Seems to be the direction the industry is going | - 92 lines and 9 params so far makes it unweidy<br />- Constructor is overloaded, handling multiple responsibilities<br />- Doesn't take advantage of OOP |
```

Comment:
> Seems to be the direction the industry is going

Please give some proof to this statement, otherwise it is hard to evaluate it.

Reply:
To name a few bigger projects
- https://github.com/boto/boto3/blob/develop/boto3/session.py#L48
- https://github.com/Miserlou/Zappa/blob/master/zappa/core.py#L243